src/mgr/DaemonServer.cc

```diff
@@ -2809,12 +2809,13 @@ void DaemonServer::got_service_map()
 	// we just started up
 	dout(10) << "got initial map e" << service_map.epoch << dendl;
 	pending_service_map = service_map;
+	pending_service_map.epoch = service_map.epoch + 1;
       } else {
 	// we we already active and therefore must have persisted it,
 	// which means ours is the same or newer.
 	dout(10) << "got updated map e" << service_map.epoch << dendl;
+	ceph_assert(pending_service_map.epoch > service_map.epoch);
```

Comment:
Nit: per the comment, shouldn't this be `>=`? I wonder if your initial issue was just due to the `pending_service_map_dirty` variable not getting updated? i.e. if the current `pending_service_map_dirty >= pending_service_map.epoch` (dirty), reset `pending_service_map_dirty` to the new epoch in this method.

Reply:
I think it should always be `pending_service_map.epoch > service_map.epoch` here.
I see only two places where `pending_service_map.epoch` is set/updated:
1) in `got_service_map` (just two lines above), if we got initial map, and then we set pending epoch to current epoch + 1
2) in `send_report`, when we advertise a new map (send the pending map to mon) and increase the pending epoch.

The comment looks correct to me if by "ours" one means the committed map, not the pending map. But I am open to suggestions how the comment could be improved.

I am not sure I understand how you propose to fix this alternatively. I assume you saw my comment in the tracker ticket [1] describing the problem scenario? I think it was wrong that we could bump  `pending_service_map.epoch` (and then `pending_service_map_dirty`) to a smaller value. They should only increase.

[1] https://tracker.ceph.com/issues/47275#note-1

Reply:
Do standby MGRs not start the daemon server? I'm fine w/ ignoring the update if it's in the past, I just feel like the assertion might be a bit extreme w/o knowing all the corner cases.

Reply:
I don't think the standby starts the daemon server.

I don't expect the corner cases, and if they really may happen I would like to know about them as soon as possible, because it may mean some other assumptions are also wrong. And the assertion will be very helpful in this case. I would prefer an mgr crash with backtrace reported instead of undefined/weird behavior.

Reply:
Fair enough -- we have plenty of time for this to soak test in teuthology.

Reply:
we've been hitting this assert intermittently in testing and most recently in the LRC when it was upgraded to pacific https://tracker.ceph.com/issues/48022